src/services/rfqm_service.ts

```diff
@@ -439,6 +450,118 @@ export class RfqmService {
+            await sleepAsync(TRANSACTION_WATCHER_SLEEP_TIME_MS);
+
+            // check if any tx has been mined
+            const receipts = await Promise.all(
+                Object.keys(submissions).map(async (transactionHash) => {
+                    return {
+                        transactionHash,
+                        response: await this._blockchainUtils.getTransactionReceiptIfExistsAsync(transactionHash),
+                    };
+                }),
+            );
+
+            for (const receipt of receipts) {
+                // if a tx has been mined, stop trying to resubmit
+                if (receipt.response !== undefined) {
+                    shouldTryResubmitting = false;
+                    const currentBlock = await this._blockchainUtils.getCurrentBlockAsync();
+                    if (receipt.response.blockNumber - currentBlock >= BLOCK_FINALITY_THRESHOLD) {
+                        isFinishedMonitoring = true;
+                    }
+                    // update all entities
+                    receipts.map((r) => {
+                        if (r.response !== undefined) {
+                            submissions[r.transactionHash].status =
+                                r.response.status === 1
+                                    ? RfqmTranasctionSubmissionStatus.Successful
+                                    : RfqmTranasctionSubmissionStatus.Reverted;
+                            submissions[r.transactionHash].blockMined = new BigNumber(r.response.blockNumber);
+                            submissions[r.transactionHash].gasUsed = new BigNumber(r.response.gasUsed);
+                        } else {
+                            submissions[r.transactionHash].status = RfqmTranasctionSubmissionStatus.DroppedAndReplaced;
+                        }
+                    });
+                    await this.dbUtils.updateRfqmTransactionSubmissionsAsync(Object.values(submissions));
+                    break;
+                }
+            }
+
+            if (shouldTryResubmitting) {
+                const newGasPrice = await this._protocolFeeUtils.getGasPriceEstimationOrThrowAsync();
+
+                if (gasPrice.lt(newGasPrice)) {
+                    gasPrice = newGasPrice;
+                    const submission = await this._submitTransactionAsync(orderHash, workerAddress, callData, gasPrice);
+                    submissions[submission.transactionHash!] = submission;
+                }
+            }
+        }
+    }
+
+    /**
+     * Determine transaction properties and submit a transaction
+     */
+    private async _submitTransactionAsync(
+        orderHash: string,
+        workerAddress: string,
+        callData: string,
+        gasPrice: BigNumber,
+    ): Promise<RfqmTransactionSubmissionEntity> {
+        // get transaction properties
+        const [nonce, gasEstimate] = await Promise.all([
+            this._blockchainUtils.getNonceAsync(workerAddress),
```

Comment:
Shouldn't we be fetching the nonce only once? `_submitTransactionAsync` is called by `completeSubmissionLifecycleAsync`. `_submitTransactionAsync`. If for some reason a transaction ends up being mined during `completeSubmissionLifecycleAsync`, after the [receipt check](https://github.com/0xProject/0x-api/pull/616/files#diff-289e53810d2a508fae0fff19710e2e2b9846a98a8935466d7e20a1490f179376R474), wouldn't we be re-submitting a new trade with a new incremented nonce?

Reply:
yes, changed this